# Modeling of Neuron Network Calculation
Calculations in neural network models involve a series of mathematical operations 
performed to process inputs, generate outputs, and perform learning on the data. 
<br><br> Note:
w1 = w1, w2 = w2, b  = θ 

In [53]:
import torch
import torch.nn.functional as F
import matplotlib.pyplot as plt

In [54]:
x1 = torch.Tensor([2.0]).double()                ; x1.requires_grad = True
x2 = torch.Tensor([0.0]).double()                ; x2.requires_grad = True
w1 = torch.Tensor([-3.0]).double()               ; w1.requires_grad = True
w2 = torch.Tensor([1.0]).double()                ; w2.requires_grad = True
b = torch.Tensor([6.8813735870195432]).double()  ; b.requires_grad = True
n = x1*w1 + x2*w2 + b 
o = torch.sigmoid(n)
cross_entropy_loss = torch.nn.functional.binary_cross_entropy(o, torch.Tensor([0.0]).double())
#loss = cross_entropy_loss(o, torch.Tensor([2.0]).double())
#loss.backward()
cross_entropy_loss.backward()

In [55]:
cross_entropy_loss

tensor(1.2279, dtype=torch.float64, grad_fn=<BinaryCrossEntropyBackward0>)

# After Checking That The Loss Is Still High Enough, We Will Replace the Variables w1, w2 and b. 
And the loss is 1.2279. We try to get the loss as close to 0 as possible. we'll find new inputs for w1, w2 and b. By subtracting each variable from the gradient of that variable.

In [56]:
w1.grad.item()

1.414213487167235

In [57]:
w2.grad.item()

0.0

In [58]:
b.grad.item()

0.7071067435836175

In [59]:
b.grad.item()

0.7071067435836175

In [60]:
w1_new = w1 - w1.grad.item()
w1_new

tensor([-4.4142], dtype=torch.float64, grad_fn=<SubBackward0>)

In [61]:
w2_new = w2 - w2.grad.item()
w2_new

tensor([1.], dtype=torch.float64, grad_fn=<SubBackward0>)

In [62]:
b_new = b - b.grad.item()
b_new

tensor([6.1743], dtype=torch.float64, grad_fn=<SubBackward0>)

# Then Replace The Variables w1, w2, and b With The New Variables.
After obtaining the new w1, w2, and b variables. Then put each variable into the model that has just been made and get a pretty good loss result of 0.0680.

In [63]:
x1 = torch.Tensor([2.0]).double()                ; x1.requires_grad = True
x2 = torch.Tensor([0.0]).double()                ; x2.requires_grad = True
w1 = torch.Tensor([-4.4142]).double()            ; w1.requires_grad = True
w2 = torch.Tensor([1.0]).double()                ; w2.requires_grad = True
b = torch.Tensor([6.1743]).double()  ; b.requires_grad = True
n = x1*w1 + x2*w2 + b 
o = torch.sigmoid(n)
cross_entropy_loss = torch.nn.functional.binary_cross_entropy(o, torch.Tensor([0.0]).double())
#loss = cross_entropy_loss(o, torch.Tensor([2.0]).double())
#loss.backward()
cross_entropy_loss.backward()

In [64]:
cross_entropy_loss

tensor(0.0680, dtype=torch.float64, grad_fn=<BinaryCrossEntropyBackward0>)